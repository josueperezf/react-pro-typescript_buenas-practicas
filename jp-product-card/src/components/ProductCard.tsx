import React, { createContext } from 'react';
import styles from '../styles/styles.module.css';

import { useProduct } from '../hooks/useProduct';
import { ProductCardContextProps, ProductCardsProps } from '../interfaces/interfaces';

// esto es para poder pasar informacion desde el componente padre al componente
export const ProductContext =  createContext({} as ProductCardContextProps);
const {Provider} = ProductContext;

export const ProductCard = ({children, product, className, style = {}, onChange, value, initialValues }: ProductCardsProps) => {
  const { counter, maxCount, cambiarValor, isMaxCountReached, reset} = useProduct({onChange, product, value, initialValues});
  return (
    <Provider value={{product, counter, cambiarValor, maxCount}}>
      <div className={`${styles.productCard} ${className}`} style={style} >
        {
          children && children({
            count: counter,
            maxCount: (initialValues && initialValues.maxCount) ? initialValues?.maxCount : undefined,
            isMaxCountReached: isMaxCountReached,
            product,
            cambiarValor,
            reset
          })
        }
      </div>
    </Provider>
  );
};