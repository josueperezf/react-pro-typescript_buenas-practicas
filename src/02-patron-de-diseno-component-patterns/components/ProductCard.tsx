import styles from '../styles/styles.module.css';

import { useProduct } from '../hooks/useProduct';
import { createContext } from 'react';
import { ProductCardContextProps, ProductCardsProps } from '../interfaces/interfaces';

// esto es para poder pasar informacion desde el componente padre al componente
export const ProductContext =  createContext({} as ProductCardContextProps);
const {Provider} = ProductContext;

export const ProductCard = ({children, product, className, style = {}, onChange, value, initialValues }: ProductCardsProps) => {
  const { counter, maxCount, cambiarValor, isMaxCountReached, reset} = useProduct({onChange, product, value, initialValues});
  return (
    <Provider value={{product, counter, cambiarValor, maxCount}}>
      <div className={`${styles.productCard} ${className}`} style={style} >
        {
          children && children({
            count: counter,
            maxCount: initialValues?.maxCount,
            isMaxCountReached: isMaxCountReached,
            product,
            cambiarValor,
            reset
          })
        }
      </div>
    </Provider>
  );
};