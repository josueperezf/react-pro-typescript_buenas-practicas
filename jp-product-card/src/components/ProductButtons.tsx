import React, { useCallback, useContext } from "react";
import { ProductButtonsProps } from "../interfaces/interfaces";
import styles from '../styles/styles.module.css';
import { ProductContext } from "./ProductCard";


export const ProductButtons = ({className = '', style}: ProductButtonsProps) => {
  console.log('renderizo');
  
  // extraer del context una nueva propiedad que se llamen maxCount 'opcional'
  /**
   * El useCallback esta optimizado para memorizar la función y nos permitiría ejecutarla de nuevo sin causar re-renders... es decir la llamamos y ya.
   * 
   * El useMemo sirme para memorizar el valor, cosa que también nos podría servir para este caso, porque memorizamos el valor si alcanzamos el valor máximo, pero si tuvieramos que ejecutar una función el useMemo no nos serviría ahí
   * 
   * Usualmente el useMemo se usa para memorizar resultados de tareas pesadas o que toman tiempo, pero en sí, para este caso es indiferente, porque ambos tendrían la mismas dependencias.
   */
    const {counter, cambiarValor, maxCount} = useContext(ProductContext);
    const isMaxReached = useCallback(() => !!maxCount && counter === maxCount , [maxCount, counter]);
    

    return (
      <div className={ `${styles.buttonsContainer} ${className}`} style={ style } >
          <button className={styles.buttonMinus} onClick={()=>cambiarValor(-1)} >-</button>
          <div className={styles.countLabel} >{counter}</div>
          {/* yo lo hice si el uso del callback, como esta comentado, y la verdad no  note cambio en el rederizado.
            segun el memo es cuando por ejemplo, tenemos tres contadores, en el mismo componente, y notamos que al cambiar uno, se redibujan todos, 
            esto hace mas lento todo, pero en este caso n o observo este  tipo de comportamientos.
           */}
          {/* <button className={`${styles.buttonAdd} ${(maxCount == counter ) && styles.disabled  } `} onClick={()=>cambiarValor(1)} >+</button> */}
          <button className={`${styles.buttonAdd} ${ isMaxReached() && styles.disabled  } `} onClick={()=>cambiarValor(1)} >+</button>
      </div>
    )
}