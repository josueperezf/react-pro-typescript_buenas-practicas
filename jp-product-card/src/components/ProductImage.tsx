import React, { useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css';
import noImage from '../assets/no-image.jpg';
import { ProductImageProps } from "../interfaces/interfaces";

export  const ProductImage = ({img = '', className = '', style}: ProductImageProps)=> {
    const {product} = useContext(ProductContext);
    const imagen = (img) ? img : product.img;

    return (<img className={`${styles.productImg} ${className} `} src={ imagen ? imagen : noImage} alt='' style={style}  />)
}