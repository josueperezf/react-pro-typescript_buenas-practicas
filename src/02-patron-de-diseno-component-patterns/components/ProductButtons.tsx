import { useContext } from "react";
import { ProductButtonsProps } from "../interfaces/interfaces";
import styles from '../styles/styles.module.css';
import { ProductContext } from "./ProductCard";

export const ProductButtons = ({className = '', style}: ProductButtonsProps) => {
    const {counter, cambiarValor} = useContext(ProductContext);
    return (
      <div className={ `${styles.buttonsContainer} ${className}`} style={ style } >
          <button className={styles.buttonMinus} onClick={()=>cambiarValor(-1)} >-</button>
          <div className={styles.countLabel} >{counter}</div>
          <button className={styles.buttonAdd} onClick={()=>cambiarValor(1)} >+</button>
      </div>
    )
}