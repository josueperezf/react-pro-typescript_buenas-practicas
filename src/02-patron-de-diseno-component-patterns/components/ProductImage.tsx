import { useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css';
import noImage from '../assets/no-image.jpg';

export  const ProductImage = ({img = ''})=> {
    const {product} = useContext(ProductContext);
    const imagen = (img) ? img : product.img;

    return (<img className={styles.productImg} src={ imagen ? imagen : noImage} alt='' />)
}