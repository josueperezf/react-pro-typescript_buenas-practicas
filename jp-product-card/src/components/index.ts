import { ProductCard as ProductCardComponent } from './ProductCard';

import { ProductTitle } from './ProductTitle';
import { ProductImage } from './ProductImage';
import { ProductButtons } from './ProductButtons';
import { ProductCardHOCProps } from '../interfaces/interfaces';

export const ProductCard: ProductCardHOCProps = Object.assign(ProductCardComponent, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductButtons
});

export {
    ProductTitle,
    ProductImage,
    ProductButtons,
}

export default ProductCard;
