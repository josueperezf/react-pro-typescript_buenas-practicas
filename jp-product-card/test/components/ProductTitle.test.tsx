import React from "react";
import  renderer from "react-test-renderer";
import { ProductTitle, ProductCard } from '../../src/components/';
import { products } from '../data/products';

describe('ProductTitle', () => {

  test('debe de mostrar el componente correctamente con el titulo esperado', () => {
    const wrapper = renderer.create(
      <ProductTitle title="titulo para prueba"/>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
  
  test('debe de mostar el componente con el nombre del producto', () => {
    const wrapper = renderer.create(
      <ProductCard product={products[0]}>
        {
          () => (
            <>
              <ProductTitle title="titulo para prueba"/>
            </>
          )
        }
      </ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
  
});
