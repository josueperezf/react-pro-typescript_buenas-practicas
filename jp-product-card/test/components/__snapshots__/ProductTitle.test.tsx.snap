// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle debe de mostar el componente con el nombre del producto 1`] = `
<div
  className="productCard undefined"
  style={Object {}}
>
  <span
    className="productDescription undefined"
  >
    titulo para prueba
     
  </span>
</div>
`;

exports[`ProductTitle debe de mostrar el componente correctamente con el titulo esperado 1`] = `
<span
  className="productDescription undefined"
>
  titulo para prueba
   
</span>
`;
